File: srm-744-div-2/second.hs
module MagicNumbersAgain (count) where

digits x = if x == 0 then [0] else reverse $ digitsRev x
  where digitsRev 0 = []
        digitsRev x = (x `mod` 10) : (digitsRev $ x `div` 10)

count :: Integer -> Integer -> Int
count a b = length $ filter (isMagicNumber . (^2)) [(ceiling $ sqrt (fromIntegral a))..(floor $ sqrt (fromIntegral b))]
  where
    isMagicNumber = areDigitsMagic True . reverse . digits
    areDigitsMagic isSmaller (d1:d2:ds) = (if isSmaller then d1 < d2 else d1 > d2) && areDigitsMagic (not isSmaller) (d2:ds)
    areDigitsMagic _ _ = True

File: srm-744-div-2/second.cpp
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

vector<int> digits(20);

class MagicNumbersAgain {
public:

    bool isMagicNumber(long long x) {
        int numDigits = 0;

        do {
            digits[numDigits++] = x % 10;
            x /= 10;
        } while (x > 0);


        for (int i = numDigits - 1; i >= 1; i--) {
            if ((i - (numDigits - 1)) % 2 == 0) {
                if (!(digits[i] < digits[i-1])) return false;
            } else {
                if (!(digits[i] > digits[i-1])) return false;
            }
        }

        return true;
    }

    int count(long long A, long long B) {
        long long a = (long long) trunc(sqrt(A));
        if (a * a < A) a += 1;

        int result = 0;
        for (long x = a; x * x <= B; x++) {
            if (isMagicNumber(x*x)) result++;
        }

        return result;
    }
};
